src/arc.rs

```diff
@@ -184,6 +176,33 @@ impl<T: ?Sized> Arc<T> {
         this.as_ptr()
     }
 
+    /// Reconstruct the `Arc<T>` from a raw pointer obtained from into_raw()
+    ///
+    /// Note: This raw pointer will be offset in the allocation and must be preceded
+    /// by the atomic count.
+    ///
+    /// It is recommended to use OffsetArc for this
+    ///
+    ///  # Safety
+    /// - The given pointer must be a valid pointer to `T` that came from [`Arc::into_raw`].
+    /// - After `from_raw`, the pointer must not be accessed.
+    #[inline]
+    pub unsafe fn from_raw(ptr: *const T) -> Self {
+        // To find the corresponding pointer to the `ArcInner` we need
+        // to subtract the offset of the `data` field from the pointer.
+
+        // SAFETY: `ptr` comes from `Arc`, so it must be initialized.
+        let offset_of_data = ArcInner::<T>::offset_of_data(ptr);
```

Comment:
I think I'd prefer an explicit documentation of the fact that `from_raw_inner` wants the beginning of the allocation, and `ptr` points to the data.